### PR TITLE
metrics exemplar round 1

### LINKED ISSUE #1093: Metrics SDK: Add Metrics exemplar
Specs: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#exemplar

- A metrics **exemplar** is a trace that is representative of the metrics being collected. The spec requires that exemplars be implemented in a way such that they may be disabled and that if disabled they should have no performance impact (collection should be disabled, not just export). When enabled, the exemplars will come from the currently active context.
_Properties_: Value, Time, Attributes, Trace ID, Span ID

- **ExemplarFilter** determines if a measurement should be sampled
_Properties_ - value of measurement, attributes of measurement, context of measurement, timestamp

- **ExemplarReservoir** provides a method to offer measurements to the reservoir and to collect accumulated exemplars

### PR DESCRIPTION
**Pull request by esigo**
Fixes #1093 (issue)

## Changes
Adds ExemplarFilter and ExemplarReservoir
missing:
Filtered, fixed size, and histogram exemplar reservoir.

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [x] Unit tests have been added
* [ ] Changes in public API reviewed

### CODE REVIEW

File: api/include/opentelemetry/metrics/sync_instruments.h [hunk 1]
```diff
@@ -104,29 +114,33 @@ class UpDownCounter : public SynchronousInstrument
...
-               attributes) noexcept
+           std::initializer_list<std::pair<nostd::string_view, common::AttributeValue>> attributes,
+           const opentelemetry::context::Context &context) noexcept
   {
-    this->Add(value, nostd::span<const std::pair<nostd::string_view, common::AttributeValue>>{
-                         attributes.begin(), attributes.end()});
+    this->Add(value,
+              nostd::span<const std::pair<nostd::string_view, common::AttributeValue>>{
+                  attributes.begin(), attributes.end()},
+              context);
```
Comment (lalitb) [inline api/include/opentelemetry/metrics/sync_instruments.h:143]: Should we make `context` optional, and use default runtime context if not provided. 

```cpp
void Add(value);
void Add(value, context);
void Add(value, attributes);
void Add(value, attributes, context);
```

or just

```cpp
void Add(value)
void Add(value, attributes)
void Add(value, attributes, context)
```

File: sdk/include/opentelemetry/sdk/metrics/examplar/filter.h [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+/**
+ * Exemplar filters are used to pre-filter measurements before attempting to store them in a
+ * reservoir.
+ */
+class ExemplarFilter
+{
+public:
+  // Returns whether or not a reservoir should attempt to filter a measurement.
+  virtual bool shouldSampleMeasurement(long value,
```
[the remark this answers is not in the document]
Comment (esigo) [reply sdk/include/opentelemetry/sdk/metrics/examplar/filter.h]: thanks done.

File: sdk/include/opentelemetry/sdk/metrics/examplar/always_sample_filter.h [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+{
+namespace metrics
+{
+
+class AlwaysSampleFilter final : public ExemplarFilter
+{
+public:
+  static nostd::shared_ptr<ExemplarFilter> GetAlwaysSampleFilter()
+  {
+    return nostd::shared_ptr<ExemplarFilter>{new AlwaysSampleFilter{}};
```
[the remark this answers is not in the document]
Comment (esigo) [reply sdk/include/opentelemetry/sdk/metrics/examplar/always_sample_filter.h]: thanks, done

File: api/include/opentelemetry/metrics/sync_instruments.h (hunk 1 above)
Comment (esigo) [reply api/include/opentelemetry/metrics/sync_instruments.h:189]: thanks, added these four:

```cpp
void Add(value);
void Add(value, context);
void Add(value, attributes);
void Add(value, attributes, context);
```

File: sdk/include/opentelemetry/sdk/metrics/examplar/reservoir.h [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * <p>This represents a reservoir for a specific "point" of metric data.
+ */
+class ExemplarReservoir
+{
+public:
+  virtual ~ExemplarReservoir() = default;
+
+  /** Offers a long measurement to be sampled. */
+  virtual void OfferMeasurement(long value,
```
Comment (lalitb) [inline sdk/include/opentelemetry/sdk/metrics/examplar/reservoir.h:25]: nit - Should this method also accept the time when the measurement was taken (as per the specs ) : https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#exemplarreservoir

File: sdk/include/opentelemetry/sdk/metrics/state/multi_metric_storage.h [hunk 5]
```diff
@@ -18,46 +19,57 @@ namespace metrics
...
   }
 
-  virtual void RecordLong(
-      long value,
-      const opentelemetry::common::KeyValueIterable &attributes) noexcept override
+  virtual void RecordLong(long value,
+                          const opentelemetry::common::KeyValueIterable &attributes,
+                          const opentelemetry::context::Context &context) noexcept override
   {
+    exemplar_reservoir_->OfferMeasurement(value, attributes, context);
```
Comment (lalitb) [inline sdk/include/opentelemetry/sdk/metrics/state/multi_metric_storage.h:42]: Should this method be called here, or just inside all the storage?

Comment (lalitb) [inline sdk/include/opentelemetry/sdk/metrics/state/multi_metric_storage.h:42]: Should this method be called here, or just inside all the storage?

File: sdk/include/opentelemetry/sdk/metrics/examplar/data.h [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+public:
+  /**
+   * The set of key/value pairs that were filtered out by the aggregator, but recorded alongside the
+   * original measurement. Only key/value pairs that were filtered out by the aggregator should be
+   * included
+   */
+  MetricAttributes GetFilteredAttributes();
+
+  /** Returns the timestamp in nanos when measurement was collected. */
+  long GetEpochNanos();
```
Comment (lalitb) [inline sdk/include/opentelemetry/sdk/metrics/examplar/data.h:32]: nit - timestamp is represented as sdk::common::SystemTimestamp at other places.

Comment (lalitb) [inline sdk/include/opentelemetry/sdk/metrics/examplar/data.h:32]: nit - timestamp is represented as sdk::common::SystemTimestamp at other places.

**Review by lalitb — APPROVED** [review]
LGTM with few minor comments.

**Review by reyang — CHANGES_REQUESTED** [review]
Need to rename "examplar" to "exemplar".

**Comment by lalitb** [comment]
> Need to rename "examplar" to "exemplar".

Good observation :)

File: sdk/include/opentelemetry/sdk/metrics/examplar/reservoir.h (hunk 4 above)
Comment (esigo) [reply sdk/include/opentelemetry/sdk/metrics/examplar/reservoir.h]: Thanks added timestamp.

File: sdk/include/opentelemetry/sdk/metrics/state/multi_metric_storage.h (hunk 5 above)
Comment (esigo) [reply sdk/include/opentelemetry/sdk/metrics/state/multi_metric_storage.h]: removed from the `MultiMetricStorage`.

File: sdk/include/opentelemetry/sdk/metrics/examplar/data.h (hunk 6 above)
Comment (esigo) [reply sdk/include/opentelemetry/sdk/metrics/examplar/data.h]: Thanks, done.

**Comment by esigo** [comment]
> Need to rename "examplar" to "exemplar".

Good catch, thanks.
Fixed the typo.

**Review by reyang — APPROVED** [review]
LGTM for round1.

**Pull request merged**